Upgrade to ipfs@0.41 (#381)
ProtoSchool/protoschool.github.io: src/tutorials/0004-mutable-file-system/10.md

The method takes this format:

```js
ipfs.files.read(path, [options])
```

The `path` provided is the path of the file to read, and it must point to a file rather than a directory.

> The `files.read` method returns an Async Iterable, that iterates over the file's chunks of data, i.e. Buffers, which can be converted to a string using the method `toString('utf8')`. For example:

Reviewer: ```suggestion
The `files.read` method returns an Async Iterable that iterates over the file's chunks of data, i.e. Buffers, which can be converted to a string using the method `toString('utf8')`. For example:
```

```js

> // the concat variable is globally available (just like ipfs)

Reviewer: We need to introduce `concat` & `it-concat` before using it.

Author: updated here: https://github.com/ProtoSchool/protoschool.github.io/pull/381/commits/771fa666432f40e0f4363d0d695bd6246db5e610

let bufferedContents = await concat(ipfs.files.read('/directory/some-file.txt'))  // a buffer
let contents = bufferedContents.toString('utf8') // a string
```

> When you're ready to try this in the real world, you should note that the above example can result in heavy memory usage, depending on the contents of the file being read. If you find this to be the case, you might want to skip the usage of the `it-concat` package and process each chunk of data iteratively.

Reviewer: ```suggestion
When you're ready to try this in the real world, you should note that the above example can result in heavy memory usage, depending on the contents of the file being read. If you're working with large files and find this to be the case, you might want to skip using the `it-concat` package and instead process each chunk of data iteratively. The main reason IPFS now returns `Async Iterables` is to provide a built-in option for dealing with potential performance issues.
```

> This is the main reason why ipfs returns `Async Iterables` so that we can decide if we want to be aware of memory performance issues or not.

Reviewer: ```suggestion
```

> In ProtoSchool tutorials, we don't really need to worry about this since our code challenges use small files, so we can concatenate everything without worrying about performance.

Reviewer: ```suggestion
In ProtoSchool tutorials, our code challenges use small files, so we can concatenate everything without worrying about performance.
```

The method takes this format:

```js
ipfs.files.read(path, [options])
```

The `path` provided is the path of the file to read, and it must point to a file rather than a directory.

The `files.read` method returns an Async Iterable, that iterates over the file's chunks of data, i.e. Buffers, which can be converted to a string using the method `toString('utf8')`. For example:

```js
const concat = require('it-concat')

let bufferedContents = await concat(ipfs.files.read('/directory/some-file.txt'))  // a buffer
let contents = bufferedContents.toString('utf8') // a string
```

or

```js
const concat = require('it-concat')

let contents = (await concat(ipfs.files.read('/directory/some-file.txt'))).toString('utf8') // a string
// notice the parentheses around the entire await statement

> ```

Reviewer 2: Personally I'd remove this second example - it's not super readable and doesn't help with much other than to save on 1 line of code.

When you're ready to try this in the real world, you should note that the above example can result in heavy memory usage, depending on the contents of the file being read. If you find this to be the case, you might want to skip the usage of the `it-concat` package and process each chunk of data iteratively.
This is the main reason why ipfs returns `Async Iterables` so that we can decide if we want to be aware of memory performance issues or not.

> But in ProtoSchool tutorials, we don't really need to worry about this since our code challenges will use small files, so we can concatenate everything without worrying about performance.

Reviewer 2: ```suggestion
In ProtoSchool tutorials, we don't really need to worry about this since our code challenges use small files, so we can concatenate everything without worrying about performance.
```

The `path` provided is the path of the file to read, and it must point to a file rather than a directory.

> The `files.read` method returns an Async Iterable that iterates over the file's chunks of data, i.e. Buffers. In our case, we ultimately need to convert Buffers into strings using the method `toString()`. However, the chunks of data within a single file need to be reassembled (concatenated) before the conversion. The [`it-to-buffer`](https://www.npmjs.com/package/it-to-buffer) package can iterate over all of the chunks and put them back together for us. (We've made this package availabe to you in our excercises as `concat`.)

Reviewer: ```suggestion
The `files.read` method returns an Async Iterable that iterates over the file's chunks of data, i.e. Buffers. In our case, we ultimately need to convert Buffers into strings using the method `toString()`. However, the chunks of data within a single file need to be reassembled (concatenated) before the conversion. The [`it-to-buffer`](https://www.npmjs.com/package/it-to-buffer) package can iterate over all of the chunks and put them back together for us. (We've made this package availabe to you in our exercises as `toBuffer`.)
```

The `path` provided is the path of the file to read, and it must point to a file rather than a directory.

The `files.read` method returns an Async Iterable that iterates over the file's chunks of data, i.e. Buffers. In our case, we ultimately need to convert Buffers into strings using the method `toString()`. However, the chunks of data within a single file need to be reassembled (concatenated) before the conversion. The [`it-concat`](https://www.npmjs.com/package/it-concat) package can iterate over all of the chunks and put them back together for us:

```js

> import concat from 'it-concat'

Reviewer: ```suggestion
// the concat function comes from the it-concat package
// and is made globally available (just like ipfs) in our code challenges
```
